lib/handler/connect.c

```diff
@@ -520,6 +536,10 @@ static int tcp_write(void *_self, int is_end_stream)
     if (self->sock != NULL && !h2o_socket_is_writing(self->sock))
         tcp_do_write(self);
 
+    if (is_end_stream) {
+        H2O_PROBE_CONN(CONNECT_TCP_WRITE_CLOSED, self->src_req->conn, self->src_req->req_index);
```

Comment:
I was thinking that it is useful to know whether it was the client or the origin who decided to close the connection, that is why there is write-closed probe and a read-closed probe.

But maybe we want to have just one probe in `close_socket` and report in a parameter what caused the close.

Reply:
> I was thinking that it is useful to know whether it was the client or the origin who decided to close the connection, that is why there is a write-closed probe and a read-closed probe.

I think that’s a reasonable argument.

My arguments are:
* The name of the probe being proposed might be confusing, because it follows the same naming convention as others, but unlike others, monitors the interaction with the connection handler (lib/httpX) rather than the socket.
* If our intent is to monitor interaction with the connection handler, maybe what we need is a set of probes for monitoring all reads and write events that happen with the connection handler, rather than just the close event on one direction.
* The interface between the generator (e.g., lib/handler/connect.c) and connection handler is not unique to lib/handler/connect.c. Probably, what we want to have is wrapper functions around that interface, and within those wrapper functions, define necessary probes. Then, we will have a generator-agnostic way of monitoring the interactions.

Reply:
If we have the `connect-dispose` probe and the `connect-tcp-read-error` probe reports the origin closing the connection, then looking at the trace we can easily determine which side initiated the connection closure:

```json
{"type":"connect-tcp-read","tid":31102,"seq":14,"conn-id":1,"req-id":1,"bytes-len":24,"time":1658536106802}
{"type":"connect-tcp-read-error","tid":31102,"seq":15,"conn-id":1,"req-id":1,"err":"socket closed by peer","time":1658536106802}
{"type":"connect-dispose","tid":31102,"seq":16,"conn-id":1,"req-id":1,"time":1658536106802}
```
vs
```json

```

Reply:
That makes sense, thank you!

If we have the `connect-dispose` probe and the `connect-tcp-read-error` probe reports the origin closing the connection, then looking at the trace we can easily determine which side initiated the connection closure.

The origin-side closed first:
```json
["... earlier log entries removed ..."]
{"type":"connect-tcp-read","tid":31102,"seq":14,"conn-id":1,"req-id":1,"bytes-len":24,"time":1658536106802}
{"type":"connect-tcp-read-error","tid":31102,"seq":15,"conn-id":1,"req-id":1,"err":"socket closed by peer","time":1658536106802}
{"type":"connect-dispose","tid":31102,"seq":16,"conn-id":1,"req-id":1,"time":1658536106802}
```

The client-side closed first:
```json
["... earlier log entries removed ..."]
{"type":"connect-tcp-write","tid":31103,"seq":31,"conn-id":2,"req-id":1,"bytes-len":24,"time":1658536296758}
{"type":"connect-dispose","tid":31103,"seq":32,"conn-id":2,"req-id":1,"time":1658536296758}
```

We do not need `connect-tcp-read-closed` and `connect-tcp-write-closed` for that.

Reply:
Probes `connect-tcp-read-closed` and `connect-tcp-write-closed` removed in 21c8f1dfa.

Reply:
The two probes removed in cd07df4.